Add notability checks for casks

SMillerDev opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This change copies over the formula audits for notability and makes them available for casks.

Fixes https://github.com/Homebrew/homebrew-cask/issues/78675

EricFromCanada commented on Library/Homebrew/cask/cmd/audit.rb:
```diff
@@ -7,8 +7,14 @@ class Audit < AbstractCommand
+          `audit` [<options>] [<cask>]
+
+          Check <cask> for Homebrew coding style violations. This should be run before
+          submitting a new casl. Will exit with a non-zero status if any errors are
```
```suggestion
          submitting a new cask. Will exit with a non-zero status if any errors are
```

vitorgalvao commented:
This is great. Thank you! It does change our notability requirements (we’re currently less strict than non-Homebrew-Cask), but I’m fine with that.

It doesn’t close Homebrew/homebrew-cask#78675, though, because that issue is about having checks for non-open-source software as well (which will be the majority).

In addition, in HBC’s case the notability check should maybe be an `audit` flag, because not all software is created equally. I’m specifically thinking of fonts, which are low maintenance, rarely updated, rarely break, and have their own repo. This would prevent most of them from being submitted (how many people fork or watch the repo of a font?)

reitermarkus commented:
Would be good to move the logic for these into helper functions so they can be shared between casks and formulae.

SMillerDev commented:
Yeah, just thought of a way to do that. I'll try that later today

vitorgalvao commented:
What does this use as the source to identify something as coming from (for example) github, and checking for notability, `homepage` or `url`?

SMillerDev commented:
> What does this use as the source to identify something as coming from (for example) github, and checking for notability, `homepage` or `url`?

Yes

vitorgalvao commented:
I meant “which of the two does it use?” Does your answer mean it uses either? So if `homepage` isn’t a github URL but `url` is, it’ll still count?

SMillerDev commented:
Yes, or appcast. Otherwise it just returns.

SMillerDev commented:
I'm not sure why this fails on `brew man`. Any suggestions?

MikeMcQuaid commented:
@SMillerDev There were some CI issues recently with this. Try to `brew update` and rebase this onto `origin/master`.

SMillerDev commented:
> In addition, in HBC’s case the notability check should maybe be an audit flag, because not all software is created equally. I’m specifically thinking of fonts, which are low maintenance, rarely updated, rarely break, and have their own repo. This would prevent most of them from being submitted (how many people fork or watch the repo of a font?)

I think something like a `--new-cask` option might be useful. I don't think you should losen the restrictions because you think some software might not make it though. If people don't have a big enough following I don't think that's on Homebrew to work around that.

vitorgalvao commented:
> If people don't have a big enough following I don't think that's on Homebrew to work around that.

Agreed. But I don’t think one can accurately measure the following of a font by number of forks. That’s an unfair metric.

It’s inconvenient that HB and HBC need different approaches or rules in many places, but we do need those differences. We serve different types of software and not everything can be evaluated the same way.

Also, [as @core-code remembered](https://github.com/Homebrew/homebrew-cask/pull/81265#issuecomment-619139464), this will cause problems in situations where an app uses GitHub only for hosting. I do frown upon that practice, but it does happen and doesn’t reflect the popularity of the software. This could be avoided if the notability check only considered `homepage`, but I understand that would mean changes to the code we want to share between projects and might fail on github-hosted pages, so there’s no perfect solution.

This is fine for now and I think we should merge it. But I also think we should all be aware of the limitations of the approach, so someone can (eventually) make an informed suggestion of a solution.

MikeMcQuaid commented:
> Agreed. But I don’t think one can accurately measure the following of a font by number of forks. That’s an unfair metric.

That's not the metric. It's a floor of the lowest number of forks, watchers or stars.

> It’s inconvenient that HB and HBC need different approaches or rules in many places, but we do need those differences. 

You _want_ those differences. You do not _need_ them. I'm not saying this is the particular case but I'm pretty disappointed by the resistance amongst the Homebrew/cask team in further unifying our efforts.

This division is perhaps why requests for @Homebrew/cask code review go ignored and I end up merging cask PRs without any real context so the contributors aren't disheartened. This division is also perhaps why we're [just deleting cask upgrade tests](https://github.com/Homebrew/brew/commits/master/Library/Homebrew/test/cask/cmd/upgrade_spec.rb) because no-one is attempting to resolve their flakiness. 

As far as I can currently tell the cask parts of Homebrew/brew are currently maintained by a single (busy) person. That gives it a bus factor of one. I'm very much not pointing the finger at that (busy) person but we need to try to bridge the cultures between these two projects if we don't want to have problems.

vitorgalvao commented:
> You _want_ those differences.

I definitely don’t. I want less effort required on our part, not more. The differences between closed-source and open-source software (and their distribution) creates difficulties in having rules that can be applied to both equally.

I just said that I think this is good and I think we should merge it. None of my comments on this thread has been blocking towards inclusion (because it should be included, it’s way better than the nothing we have now). I like *and asked for* this.

> This division is perhaps why requests for @Homebrew/cask code review go ignored and I end up merging cask PRs without any real context so the contributors aren't disheartened. This division is also perhaps why we're [just deleting cask upgrade tests](https://github.com/Homebrew/brew/commits/master/Library/Homebrew/test/cask/cmd/upgrade_spec.rb) because no-one is attempting to resolve their flakiness.
> 
> As far as I can currently tell the cask parts of Homebrew/brew are currently maintained by a single (busy) person. That gives it a bus factor of one. I'm very much not pointing the finger at that (busy) person but we need to try to bridge the cultures between these two projects if we don't want to have problems.

But on that, I fully agree with the criticism and it has been worrying me for a good while. I do think we need more people working on the core. @vladimyr has recently caught my eye as a good contributor (which even had an addition merged to the core already), and I do appreciate that @SMillerDev took an interest as well. Also, I just got word from another contributor relating to more work on the core.

I also agree it’s unfair you have to be the one to merge those PRs. I wish I knew sooner of your feelings regarding that (maybe you’ve expressed it before and I missed it).

reitermarkus commented:
> I think something like a `--new-cask` option might be useful.

I think that's the last thing missing here.

@SMillerDev, have you checked how many casks are failing this check currently?

reitermarkus commented on Library/Homebrew/cask/audit.rb:
```diff
@@ -322,6 +326,53 @@ def check_appcast_contains_version
+      odebug "Auditing GitHub repo"
+
+      warning = SharedAudits.github(user, repo)
+      return if warning.nil?
```
Maybe simplify these to

```ruby
add_error warning if warning
```

and maybe also rename `warning` to `error`.

MikeMcQuaid commented:
> I just said that I think this is good and I think we should merge it. 

Apologies, I should have left it given that.

> But on that, I fully agree with the criticism and it has been worrying me for a good while. I do think we need more people working on the core.

❤️ 

> I also agree it’s unfair you have to be the one to merge those PRs. I wish I knew sooner of your feelings regarding that (maybe you’ve expressed it before and I missed it).

It's unfair of me to not raise it publicly rather than write a badly worded comment in an unrelated thread. I've expressed it with hints here and there but that's no way to handle communication of this sort. Again: I apologise @vitorgalvao and to the other cask folks for both handling and wording this badly.

reitermarkus commented:
I think we should remove the check for the `appcast`, since most casks which have e.g. a GitHub appcast will also have the `url` and/or `homepage` on GitHub as well.

SMillerDev commented:
Updated but forgot to remove the appcast fallback. It might not be used much but I don't think it'll hurt.

I also tried to clean up the arguments to methods when I added the new ones, if you need me to split the commands into a seperate PR let me know.

reitermarkus commented:
> Updated but forgot to remove the appcast fallback. It might not be used much but I don't think it'll hurt.

Have you checked how many (if any) casks are failing this check currently?

SMillerDev commented on Library/Homebrew/cask/cmd/audit.rb:
```diff
@@ -1,32 +1,66 @@
 
       def self.help
-        "verifies installability of Casks"
+        usage_banner
```
It's not super pretty but it's not broken either.

reitermarkus replied:
> It's not super pretty

Yeah, that's what I meant by broken output. All others help texts are a single line, this text spans over multiple.

SMillerDev commented:
> Have you checked how many (if any) casks are failing this check currently?

There's a 132 casks currently failing the repo checks, most of them on notability. I'm checking without appcast now.

SMillerDev commented:
There's 123 casks failing the repo check if you disable appcast checking. That means 9 failing ones are only caught on the appcast.

core-code commented:
> That means 9 failing ones are only caught on the appcast.

its expected that some of them fail as appcasts are sometimes outdated. i'm surprised it is so few, last time i checked it was at least a hundred.

SMillerDev commented:
> its expected that some of them fail as appcasts are sometimes outdated. i'm surprised it is so few, last time i checked it was at least a hundred.

I'm refering to them failing the notability check. The value of the appcast is irrelevant there.

core-code commented:
ah ok that explains it :)

reitermarkus commented:
> That means 9 failing ones are only caught on the appcast.

Are those 9 failures true positives?

SMillerDev commented:
Build failure is in the docker image generation.

SMillerDev commented:
Should I merge this myself?

MikeMcQuaid commented:
@SMillerDev yup 😁
